lib-nearby/src/main/java/com/meshrabiya/lib_nearby/nearby/NearbyVirtualNetwork.kt

```diff
@@ -0,0 +1,351 @@
+                log("Failed to request connection to endpoint: $endpointId", e)
+                updateEndpointStatus(endpointId, EndpointStatus.DISCONNECTED)
+            }
+    }
+
+    private fun log(message: String, exception: Exception? = null) {
+        val prefix = "[NearbyVirtualNetwork:$name] "
+        if (exception != null) {
+            logger(Log.ERROR, "$prefix$message", exception)
+        } else {
+            logger(Log.DEBUG, "$prefix$message")
+        }
+    }
+
+    private fun observeEndpointStatusFlow() {
+        CoroutineScope(Dispatchers.Default).launch {
+            endpointStatusFlow.collect { endpointMap ->
+                checkAndInitiateConnections(endpointMap)
+            }
+        }
+    }
+
+    private fun checkAndInitiateConnections(endpointMap: Map<String, EndpointInfo>) {
+        val connectedOutgoing = endpointMap.values.count { it.status == EndpointStatus.CONNECTED && it.isOutgoing }
+        val disconnectedEndpoints = endpointMap.values.filter { it.status == EndpointStatus.DISCONNECTED }
+
+        log( "Connected outgoing: $connectedOutgoing, Disconnected endpoints: ${disconnectedEndpoints.size}")
+
+        if (connectedOutgoing < desiredOutgoingConnections && disconnectedEndpoints.isNotEmpty()) {
+            val endpointsToConnect = disconnectedEndpoints.take(desiredOutgoingConnections - connectedOutgoing)
+            log( "Initiating connection to ${endpointsToConnect.size} endpoints")
+            updateMultipleEndpointsStatus(endpointsToConnect.map { it.endpointId }, EndpointStatus.CONNECTING)
+            endpointsToConnect.forEach { requestConnection(it.endpointId) }
+        }
+    }
+
+    private fun updateMultipleEndpointsStatus(endpointIds: List<String>, status: EndpointStatus) {
+        _endpointStatusFlow.update { currentMap ->
+            currentMap.toMutableMap().apply {
+                endpointIds.forEach { endpointId ->
+                    this[endpointId]?.status = status
+                    this[endpointId]?.isOutgoing = true
+                }
+            }
+        }
+    }
+
+    private fun sendMmcpPingPacket(endpointId: String) {
+        val mmcpPing = MmcpPing(Random.nextInt())
+        val pingBytes = mmcpPing.toBytes()
+
+        val header = VirtualPacketHeader(
+            fromAddr = InetAddress.getByName(virtualIpAddress).address.ip4AddressToInt(),
+            toAddr = InetAddress.getByName(broadcastAddress).address.ip4AddressToInt(),
+            fromPort = 0,
+            toPort = 0,
+            lastHopAddr = InetAddress.getByName(virtualIpAddress).address.ip4AddressToInt(),
+            payloadSize = pingBytes.size,
+            hopCount = 0,
+            maxHops = 1
+        )
+
+        val totalSize = VirtualPacketHeader.HEADER_SIZE + pingBytes.size
+        val data = ByteArray(totalSize)
+        header.toBytes(data, 0)
+        System.arraycopy(pingBytes, 0, data, VirtualPacketHeader.HEADER_SIZE, pingBytes.size)
```

Comment:
No need to make a copy of the packet. That does have some (small) overhead, but as many packets are being exchanged on a regular basis, better to keep that overhead to a minimum.